rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,674 @@
+checking is, by construction, performed **against the same condition** that is
+later used as stub, which ensures soundness (see discussion on lingering threats
+to soundness in the [future](#future-possibilities) section) and guarding against
+stubs diverging from their checks.
+
+[^side-effects]: Code used in contracts is required to be side effect free which
+    means it must not perform I/O, mutate memory (`&mut` vars and such) or
+    (de)allocate heap memory. This is enforced by the verifier, see the
+    discussion in the [future](#future-possibilities) section.
+
+### Write Sets and havocking
+
+A return value is only one way in which a function may communicate data. It can
+also communicate data by modifying values stored behind mutable pointers.
+
+To simulate all possible modifications a function could apply to pointed-to data
+the verifier "havocs" those regions, essentially replacing their content with
+non-deterministic values.
+
+Let us consider a simple example of a `pop` method.
+
+```rs
+impl<T> Vec<T> {
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+This function can, in theory, modify any memory behind `&mut self`, so this is
+what Kani will assume it does by default. It infers the "write set", that is the
+set of memory locations a function may modify, from the type system. As a result
+any data pointed to by a mutable reference or pointer is considered part of the
+write set. In addition a static analysis of the source code discovers any
+`static mut` variables the function or it's dependencies reference and add all
+pointed-to data to the write set also. During havocking all locations in the
+write set are replaced by non-deterministic values by the verifier.
+
+While the inferred write set is sound and enough for successful contract
+checking[^inferred-footprint] in many cases (such as `pop`) this inference is
+too coarse grained. In this case every value in this vector will be made
+non-deterministic.
+
+This proposal also adds an `modifies` and `frees` clause which limits the scope
+of havocking. Both clauses represent an assertion that the function will modify
+only the specified memory regions. Similar to requires/ensures the verifier
+enforces the assertion in the checking stage to ensure soundness. When the
+contract is used as a stub the modifies clause is used as the write set to havoc.
+
+In our `pop` example the only modified memory location is the last element and
+only if the vector was not already empty, which would be specified thusly.
+
+```rs
+impl<T> Vec<T> {
+  #[modifies(!self.is_empty(), (*self).buf.ptr.pointer.pointer[self.len])]
+  #[modifies(self.is_empty(), false)]
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+The `#[modifies(CONDITION, MODIFIES_RANGE)]` consists of an optional `CONDITION`
+(defaults to `true`) and a `MODIFIES_RANGE` which is essentially an lvalue. 
+
+Lvalues are simple expressions permissible on the left hand side of an
```

Comment:
Your example says `#[modifies(self.is_empty(), false)]` but `false` isn't an lvalue?

Reply:
Good catch thanks. I've revised the syntax to be able to express this without `false`